File: packages/sdk/src/commands/validate.ts
// validate command — WD-5. Parse SKILL.md frontmatter and validate it against
// the windy.drop.v1 schema. Print clear errors with path information.
//
// Exit codes:
//   0 — manifest is valid
//   1 — manifest is invalid (or SKILL.md missing / malformed)

import { DropManifestSchema } from "@windy/drops-artifact-spec";
import chalk from "../lib/chalk-shim.js";
import { readSkillMd, SkillMdError } from "../lib/skill-md.js";

export interface ValidateOptions {
  path: string;
}

export interface ValidateResult {
  valid: boolean;
  errors: ValidateError[];
}

export interface ValidateError {
  path: string;       // dotted field path, e.g., "author.0.passport"
  message: string;
  filePath: string;
}

export async function validate(opts: ValidateOptions): Promise<ValidateResult> {
  let parsed;
  try {
    parsed = readSkillMd(opts.path);
  } catch (e) {
    if (e instanceof SkillMdError) {
      return {
        valid: false,
        errors: [{ path: "", message: e.message, filePath: e.filePath }],
      };
    }
    throw e;
  }

  const result = DropManifestSchema.safeParse(parsed.frontmatter);
  if (result.success) {
    return { valid: true, errors: [] };
  }

  const errors: ValidateError[] = result.error.issues.map((issue) => ({
    path: issue.path.map(String).join(".") || "(root)",
    message: issue.message,
    filePath: parsed.filePath,
  }));

  return { valid: false, errors };
}

export async function run(opts: ValidateOptions): Promise<void> {
  const result = await validate(opts);

  if (result.valid) {
    process.stdout.write(`${chalk.green("✓")} ${opts.path}/SKILL.md is valid\n`);
    process.exit(0);
  }

  for (const err of result.errors) {
    process.stderr.write(
      `${chalk.red("✗")} ${err.filePath}: ${err.path}: ${err.message}\n`,
    );
  }
  process.exit(1);
}

File: python/sdk/src/windy_drops/commands/validate.py
"""validate command — WD-5. Parse SKILL.md frontmatter and validate against
the windy.drop.v1 schema (via windy_drops_spec Pydantic models)."""

from __future__ import annotations

import sys
from dataclasses import dataclass

import typer
from pydantic import ValidationError
from windy_drops_spec import DropManifest

from ..lib.skill_md import SkillMdError, read_skill_md


@dataclass(frozen=True)
class ValidateError:
    path: str
    message: str
    file_path: str


@dataclass(frozen=True)
class ValidateResult:
    valid: bool
    errors: list[ValidateError]


def validate(*, path: str) -> ValidateResult:
    try:
        parsed = read_skill_md(path)
    except SkillMdError as e:
        return ValidateResult(
            valid=False,
            errors=[ValidateError(path="", message=str(e), file_path=e.file_path)],
        )

    try:
        DropManifest.model_validate(parsed.frontmatter)
    except ValidationError as e:
        errors = [
            ValidateError(
                path=".".join(str(p) for p in issue["loc"]) or "(root)",
                message=issue["msg"],
                file_path=str(parsed.file_path),
            )
            for issue in e.errors()
        ]
        return ValidateResult(valid=False, errors=errors)

    return ValidateResult(valid=True, errors=[])


def run(
    path: str = typer.Argument(..., help="Drop directory"),
) -> None:
    """Validate a drop's SKILL.md frontmatter against the schema."""
    result = validate(path=path)
    if result.valid:
        typer.echo(f"✓ {path}/SKILL.md is valid")
        raise typer.Exit(code=0)

    for err in result.errors:
        typer.echo(f"✗ {err.file_path}: {err.path}: {err.message}", err=True)
    raise typer.Exit(code=1)
